### PR TITLE
CSS cursor should not eagerly evaluate calc() values

### PR DESCRIPTION
**Pull request by weinig**
#### cef722bf55c74c09ac0fc5eec4a35674ab09b977
<pre>
CSS cursor should not eagerly evaluate calc() values
<a href="https://bugs.webkit.org/show_bug.cgi?id=279319">https://bugs.webkit.org/show_bug.cgi?id=279319</a>

Reviewed by Darin Adler.

Replace eager evaluation of calc() for the x/y coordinates
in CSS cursor values with normal storage or primitive values
and resolution at style building.

Retained existing casting from double -&gt; int rather than
rounding to minimize changes.

* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/parsing/cursor-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/parsing/cursor-valid.html:
    - Add tests showing retaining calc for specified value.

* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/style/StyleBuilderCustom.h:

Canonical link: <a href="https://commits.webkit.org/283316@main">https://commits.webkit.org/283316@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/661ff03dc02e0fcb32e9c138479ac7e8dee67850

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45336 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18582 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69992 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16572 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16854 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52941 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11522 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41840 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33576 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38511 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14481 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15448 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14828 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71695 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9918 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14242 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60255 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9950 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57166 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60545 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14536 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8193 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1828 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41144 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42220 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43403 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41964 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/CSSCursorImageValue.h [hunk 1]
```diff
@@ -28,19 +28,18 @@
 namespace WebCore {
 
 class StyleImage;
+class StyleCursorImage;
```
Comment (weinig) [inline Source/WebCore/css/CSSCursorImageValue.h:30]: Blerg. I need to sort these.

File: Source/WebCore/css/CSSCursorImageValue.h [hunk 2]
```diff
@@ -35,12 +36,10 @@ class BuilderState;
 
 class CSSCursorImageValue final : public CSSValue {
 public:
-    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, const std::optional<IntPoint>& hotSpot, LoadedFromOpaqueSource);
-    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, const std::optional<IntPoint>& hotSpot, URL, LoadedFromOpaqueSource);
+    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, RefPtr<CSSValue>&& hotSpot, LoadedFromOpaqueSource);
+    static Ref<CSSCursorImageValue> create(Ref<CSSValue>&& imageValue, RefPtr<CSSValue>&& hotSpot, URL, LoadedFromOpaqueSource);
```
Comment (darinadler) [inline Source/WebCore/css/CSSCursorImageValue.h:40]: I wish we had a way to express that this hot spot value is known to be a non-coalescing pair of numbers: In the type system or even with comments and assertions. For example, should this at least be `CSSValuePair` instead of just `CSSValue`?

What I don’t like is that the parser has the checking responsibility, but this code has the resolution responsibility that relies on that checking. And the computed style code has "checking" responsibility for making sure it creates the same type of object that also has to work with the resolution code. I would like to make it likely that future programmers won’t accidentally use a type other than a non-coalescing pair of `CSSPrimitiveValue` numbers, and will immediately understand they made a mistake even if there isn’t thorough test coverage.

File: Source/WebCore/style/StyleBuilderCustom.h [hunk 3]
```diff
@@ -1145,7 +1145,9 @@ inline void BuilderCustom::applyValueCursor(BuilderState& builderState, CSSValue
     auto& list = downcast<CSSValueList>(value);
     for (auto& item : list) {
         if (auto* image = dynamicDowncast<CSSCursorImageValue>(item)) {
-            builderState.style().addCursor(builderState.createStyleImage(*image), image->hotSpot());
+            auto styleImage = image->createStyleImage(builderState);
+            auto hotSpot = styleImage->hotSpot();
+            builderState.style().addCursor(WTFMove(styleImage), WTFMove(hotSpot));
```
Comment (darinadler) [inline Source/WebCore/style/StyleBuilderCustom.h]: I don’t think we need to `WTFMove` a `std::optional<IntPoint>`. It does no harm, but for brevity maybe not?

File: Source/WebCore/css/CSSCursorImageValue.cpp [hunk 4]
```diff
@@ -61,23 +62,33 @@ CSSCursorImageValue::~CSSCursorImageValue() = default;
...
+RefPtr<StyleCursorImage> CSSCursorImageValue::createStyleImage(Style::BuilderState& state) const
 {
     auto styleImage = state.createStyleImage(m_imageValue.get());
     if (!styleImage)
         return nullptr;
-    return StyleCursorImage::create(styleImage.releaseNonNull(), m_hotSpot, m_originalURL, m_loadedFromOpaqueSource);
+
+    std::optional<IntPoint> hotSpot;
+    if (m_hotSpot) {
+        // FIXME: Should we clamp or round instead of just casting from double to int?
```
Comment (darinadler) [inline Source/WebCore/css/CSSCursorImageValue.cpp:85]: Sure would be neat to fix this as a drive-by if we can figure out what is the best behavior.

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/css/CSSCursorImageValue.cpp (hunk 4 above)
Comment (weinig) [reply Source/WebCore/css/CSSCursorImageValue.cpp:85]: Yeah, seems like rounding would be ideal, but I wanted to do that separately just incase it breaks something and needs to be rolled out.

Comment (darinadler) [reply Source/WebCore/css/CSSCursorImageValue.cpp:85]: To be fair, it’s not entirely trivial because I realize there are 3 orthogonal small questions:

1) Should we round instead of truncating?
2) Should we clamp to the integer range instead of accepting the behavior that casting an out of range value from `double` to `int` achieves (can’t remember what that is)?
3) Should we clamp further based on rules about what a valid hot spot is? Apparently this is a WebKit difference from the other browser engines: https://bugs.webkit.org/show_bug.cgi?id=102398 — I also suspect that unlike (1) and (2) this would be done after style resolution, not here.

But I still think these are things you could handle if you are motivated enough.
